src/bdecode.cpp

```diff
@@ -289,6 +289,93 @@ namespace {
 		m_buffer = buf;
 	}
 
+	bool bdecode_node::has_soft_error(span<char> error) const
+	{
+		bdecode_token const* tokens = m_root_tokens;
+		int token = m_token_idx;
+
+		// we don't know what the original depth_limit was
+		// so this has to go on the heap
+		std::vector<int> stack;
+		// make the initial allocation the default depth_limit
+		stack.reserve(100);
+
+		do
+		{
+			switch (tokens[token].type)
+			{
+			case bdecode_token::integer:
+				if (m_buffer[tokens[token].offset + 1] == '0'
+					&& m_buffer[tokens[token].offset + 2] != 'e')
+				{
+					std::snprintf(error.data(), error.size(), "leading zero in integer");
+					return true;
+				}
+				break;
+			case bdecode_token::string:
+				if (m_buffer[tokens[token].offset] == '0'
+					&& m_buffer[tokens[token].offset + 1] != ':')
+				{
+					std::snprintf(error.data(), error.size(), "leading zero in string length");
+					return true;
+				}
+				break;
+			case bdecode_token::dict:
+			case bdecode_token::list:
+				stack.push_back(token);
+				break;
+			case bdecode_token::end:
+				auto const parent = stack.back();
```

Comment:
hm, I'm surprised it's OK to declare a new variable here, without enclosing braces.